src/slicer.cpp

```diff
@@ -82,122 +84,305 @@ int SlicerLayer::getNextSegmentIdx(const Mesh* mesh, const SlicerSegment& segmen
-            Point diff = open_polyline.back() - open_polyline_other[0];
+            // in order if using back of line 1 and front of line 2
+            return (terminus_0_idx & 1) &&
+                !(terminus_1_idx & 1);
+        }
 
-            if (shorterThen(diff, largest_neglected_gap_second_phase))
+        // priority_queue will give greatest first so greatest
+        // must be most desirable stitch
+        bool operator<(const PossibleStitch &other) const
+        {
+            if (dist2 > other.dist2)
             {
-                if (open_polyline_idx == open_polyline_other_idx)
-                {
-                    polygons.add(open_polyline);
-                    open_polyline.clear();
-                    break;
-                }
-                else
-                {
-                    for (unsigned int line_idx = 0; line_idx < open_polyline_other.size(); line_idx++)
-                    {
-                        open_polyline.add(open_polyline_other[line_idx]);
-                    }
-                    open_polyline_other.clear();
-                }
+                return true;
+            }
+            else if (dist2 < other.dist2)
+            {
+                return false;
+            }
+
+            // use in order connections before reverse connections
+            if (!in_order() && other.in_order())
+            {
+                return true;
+            }
+
+            if (terminus_0_idx > other.terminus_0_idx)
+            {
+                return true;
+            }
+            else if (terminus_0_idx < other.terminus_0_idx)
+            {
+                return false;
+            }
+
+            if (terminus_1_idx > other.terminus_1_idx)
+            {
+                return true;
             }
+            else if (terminus_1_idx < other.terminus_1_idx)
+            {
+                return false;
+            }
+
+            return false;
         }
-    }
-}
+    };
+
+    std::priority_queue<PossibleStitch> stitch_queue;
 
-void SlicerLayer::stitch(Polygons& open_polylines)
-{ // TODO This is an inefficient implementation which can run in O(n^3) time.
-    // below code closes smallest gaps first
-    while(1)
     {
-        int64_t best_dist2 = max_stitch1 * max_stitch1;
-        unsigned int best_polyline_1_idx = -1;
-        unsigned int best_polyline_2_idx = -1;
-        bool reversed = false;
+        int64_t max_dist2 = max_dist * max_dist;
+
+        struct StitchGridVal
+        {
+            unsigned int polyline_idx;
+            Point polyline_end;
+        };
+
+        struct StitchGridValPointAccess
+        {
+            Point operator()(const StitchGridVal &val) const
+            {
+                return val.polyline_end;
+            }
+        };
+
+        SparseGrid<StitchGridVal,StitchGridValPointAccess> grid(cell_size);
+
+        // populate grid
+        for(unsigned int polyline_0_idx = 0; polyline_0_idx < open_polylines.size(); polyline_0_idx++)
```

Comment:
The implementation uses 0 for the polyline that is the first part of the stitched polyline and 1 for the polyline that is the second part of the stitched polyline.  This is the nominal order although when reversing segments it can use the other order.